src/state/opamState.ml

```diff
@@ -1478,67 +1511,242 @@ let string_of_cnf string_of_atom cnf =
 let string_of_conjunction string_of_atom c =
   Printf.sprintf "%s" (OpamStd.List.concat_map " , " string_of_atom (List.rev c))
 
+(* true if the opam package is available on the given switch *)
+let available_filter t opam switch =
+  let t = with_switch switch t in
+  let ct = get_switch t switch in
+  let env v = 
+    if not (OpamVariable.Full.is_global v) then None else
+    match OpamVariable.to_string (OpamVariable.Full.variable v) with
+    | "ocaml-version" -> string (OpamCompiler.Version.to_string
+                                   (Lazy.force ct.compiler_version))
+    | "opam-version"  -> string (OpamVersion.to_string OpamVersion.current)
+    | "compiler" -> string (OpamCompiler.to_string ct.compiler)
+    | "preinstalled"  -> bool (OpamFile.Comp.preinstalled
+                                 (compiler_comp t ct.compiler))
+    | "switch"        -> string (OpamSwitch.to_string ct.switch)
+    | "ocaml-native"  ->
+      if ct.compiler = OpamCompiler.system then
+        bool (Lazy.force OpamOCaml.ocaml_native_available)
+      else
+        bool (OpamFilename.exists
+                (OpamPath.Switch.bin t.root ct.switch ct.switch_config
+                 // "ocamlopt"))
+    | "ocaml-native-tools" ->
+      if ct.compiler = OpamCompiler.system then
+        bool (Lazy.force OpamOCaml.ocaml_opt_available)
+      else
+        bool (OpamFilename.exists
+                (OpamPath.Switch.bin t.root ct.switch ct.switch_config
+                 // "ocamlc.opt"))
+    | "ocaml-native-dynlink" ->
+      if ct.compiler = OpamCompiler.system then
+        bool (Lazy.force OpamOCaml.ocaml_natdynlink_available)
+      else
+        bool (OpamFilename.exists
+                (OpamPath.Switch.lib_dir t.root ct.switch ct.switch_config
+                 / "ocaml" // "dynlink.cmxa"))
+    | "arch"          -> string (OpamStd.Sys.arch ())
+    | _               -> None
+  in
+  OpamFilter.eval_to_bool ~default:true env (OpamFile.OPAM.available opam)
+  && consistent_ocaml_version ct opam
+
+let pef_package ?(orphans=OpamPackage.Set.empty) switches t =
+  let depends   = OpamPackage.Map.map OpamFile.OPAM.depends t.opams in
+  let depopts   = OpamPackage.Map.map OpamFile.OPAM.depopts t.opams in
+  let conflicts = OpamPackage.Map.map OpamFile.OPAM.conflicts t.opams in
+  let maintainers = OpamPackage.Map.map OpamFile.OPAM.maintainer t.opams in
```

Comment:
My thought was that you could loop through packages and, for each package, lookup once in t.opams directly and get all of depends, depopts, conflicts and maintainers ; rather than using specific maps and looking up each package in each of them. But maybe did I miss something ?